contracts/pre-propose/cwd-pre-propose-approver/src/contract.rs

```diff
@@ -0,0 +1,272 @@
+    let propose_messsage = SubMsg::reply_on_success(
+        WasmMsg::Execute {
+            contract_addr: proposal_module.into_string(),
+            msg: to_binary(&sanitized_msg)?,
+            funds: vec![],
+        },
+        APPROVER_PROPOSE_REPLY_ID,
+    );
+
+    // Save current pre-propose-id for use in submessage reply
+    CURRENT_PRE_PROPOSE_ID.save(deps.storage, &pre_propose_id)?;
+
+    Ok(Response::default().add_submessage(propose_messsage))
+}
+
+pub fn execute_proposal_completed(
+    deps: DepsMut,
+    info: MessageInfo,
+    proposal_id: u64,
+    new_status: Status,
+) -> Result<Response, PreProposeError> {
+    // Safety check, this message can only come from the proposal module
+    let proposal_module = PrePropose::default().proposal_module.load(deps.storage)?;
+    if info.sender != proposal_module {
+        return Err(PreProposeError::NotModule {});
+    }
+
+    // Get approval pre-propose id
+    let pre_propose_id = PROPOSAL_IDS.load(deps.storage, proposal_id)?;
+
+    // Get approval contract address
+    let approval_contract = PRE_PROPOSE_APPROVAL_CONTRACT.load(deps.storage)?;
+
+    // TODO should we only handle passed and rejected?
+    if new_status != Status::Closed && new_status != Status::Executed {
+        return Err(PreProposeError::NotClosedOrExecuted { status: new_status });
+    }
+
+    // On completion send rejection or approval message
+    let msg = match new_status {
+        Status::Closed => Some(WasmMsg::Execute {
+            contract_addr: approval_contract.into_string(),
+            msg: to_binary(&PreProposeApprovalExecuteMsg::Extension {
+                msg: ApprovalExt::Reject { id: pre_propose_id },
+            })?,
+            funds: vec![],
+        }),
+        Status::Executed => Some(WasmMsg::Execute {
+            contract_addr: approval_contract.into_string(),
+            msg: to_binary(&PreProposeApprovalExecuteMsg::Extension {
+                msg: ApprovalExt::Approve { id: pre_propose_id },
+            })?,
+            funds: vec![],
+        }),
+        Status::Passed => Some(WasmMsg::Execute {
+            contract_addr: approval_contract.into_string(),
+            msg: to_binary(&PreProposeApprovalExecuteMsg::Extension {
+                msg: ApprovalExt::Reject { id: pre_propose_id },
+            })?,
+            funds: vec![],
+        }),
+        Status::Rejected => Some(WasmMsg::Execute {
+            contract_addr: approval_contract.into_string(),
+            msg: to_binary(&PreProposeApprovalExecuteMsg::Extension {
+                msg: ApprovalExt::Approve { id: pre_propose_id },
+            })?,
+            funds: vec![],
+        }),
+        // TODO Maybe don't default reject?
+        _ => None,
+    };
+
+    // TODO two match expressions annoys me
+    match msg {
+        Some(msg) => Ok(Response::default().add_message(msg)),
+        // TODO Not sure what we want to do here...
+        None => Err(PreProposeError::Unauthorized {}),
```

Comment:
this should be unreachable. the status you don't check is `ExecutionFailed` which will never actually fire as a hook. to reach that status:

1. proposal is executed
2. executed status hook fires
3. if proposal execution fails, execution failed status is set

this may be unintuitive though. what would you expect in that case as a contract author?

Reply:
we could modify that hook logic so that only one of `executed` or `execution_failed` is fired. that may be the way here.. if you agree i can do that :)

Reply:
> we could modify that hook logic so that only one of executed or execution_failed is fired. that may be the way here.. if you agree i can do that :)

Feel free to make commits or PRs into this one. Let's make it nice. : )